### PR TITLE
backport: merge bitcoin#20222, #18766, #14033, #19486, #19607, #20291, partial #20187, #20228 (deglobalize addrman and feeEstimator)

### PR DESCRIPTION
**Pull request by kittywhiskers**
Depends on https://github.com/dashpay/dash/pull/5169

### CODE REVIEW

File: src/init.cpp [hunk 1]
```diff
@@ -286,17 +283,8 @@ void PrepareShutdown(NodeContext& node)
         DumpMempool(*node.mempool);
     }
 
-    if (fFeeEstimatesInitialized)
```
Comment (knst) [inline src/init.cpp:289]: it should be in backported with 18766?

File: test/functional/feature_fee_estimation.py [hunk 2]
```diff
@@ -243,5 +244,11 @@ def run_test(self):
         self.log.info("Final estimates after emptying mempools")
         check_estimates(self.nodes[1], self.fees_per_kb)
 
+        self.log.info("Testing that fee estimation is disabled in blocksonly.")
```
Comment (knst) [inline test/functional/feature_fee_estimation.py:247]: 18766 too?

File: src/net_processing.cpp [hunk 3]
```diff
@@ -939,7 +936,7 @@ void PeerLogicValidation::FinalizeNode(const CNode& node, bool& fUpdateConnectio
     if (state->fSyncStarted)
         nSyncStarted--;
 
-    if (state->nMisbehavior == 0 && state->fCurrentlyConnected /* && !node.IsBlockOnlyConn() */) {
+    if (state->nMisbehavior == 0 && node.fSuccessfullyConnected /* && !node.IsBlockOnlyConn() */) {
```
Comment (knst) [inline src/net_processing.cpp]: original 20291 also changes order nMisbehaviour <> fSucessfullyConnected

File: src/net_processing.cpp [hunk 4]
```diff
@@ -2785,7 +2782,6 @@ void PeerLogicValidation::ProcessMessage(
         if (!pfrom.fInbound) {
             // Mark this node as currently connected, so we update its timestamp later.
             LOCK(cs_main);
```
Comment (knst) [inline src/net_processing.cpp:2831]: hm, cs_main also removed in that backport and one more line above

File: src/test/fuzz/connman.cpp [hunk 5]
```diff
@@ -23,116 +23,94 @@ void initialize_connman()
...
-    CService random_service;
     CSubNet random_subnet;
     std::string random_string;
     while (fuzzed_data_provider.ConsumeBool()) {
-        switch (fuzzed_data_provider.ConsumeIntegralInRange<int>(0, 28)) {
+        switch (fuzzed_data_provider.ConsumeIntegralInRange<int>(0, 23)) {
         case 0:
-            random_address = ConsumeAddress(fuzzed_data_provider);
-            break;
-        case 1:
```
Comment (knst) [inline src/test/fuzz/connman.cpp:38]: nit: may be better to do bitcoin#20828 before for simplicity?
    Merge #20828: fuzz: Introduce CallOneOf helper to replace switch-case

**Review by knst — CHANGES_REQUESTED** [review]
20222 - seems as there is a mixture of changes from 20222 and from 18766. 
Better to do as spare commit. So far it looks for me as a merge/conflict mistake when 18766 and 20222 squashed together by mistakes.

20187 - seems as big missing chunk (should be marked partial then):
```
        if (!pfrom.IsInboundConn()) {
            // For non-inbound connections, we update the addrman to record
            // connection success so that addrman will have an up-to-date
            // notion of which peers are online and available.
            //
            // While we strive to not leak information about block-relay-only
            // connections via the addrman, not moving an address to the tried
            // table is also potentially detrimental because new-table entries
            // are subject to eviction in the event of addrman collisions.  We
            // mitigate the information-leak by never calling
            // CAddrMan::Connected() on block-relay-only peers; see
            // FinalizeNode().
            //
            // This moves an address from New to Tried table in Addrman,
            // resolves tried-table collisions, etc. 
```            
I recommend to do bitcoin#19724 (and may be some others related) prior this backport 20187.
    Merge #19724: [net] Cleanup connection types- followups
 



 #20228 - partial because missing changes from rpc/net.cpp
 That rpc would be introduced only in
     Merge #19658: [rpc] Allow RPC to fetch all addrman records and add records to addrman
 
Also after bitcoin#20828 would be much easier to backport src/test/fuzz/connman.cpp

I feel like these PR is ahead of time ATM

File: src/test/fuzz/connman.cpp (hunk 5 above)
Comment (kittywhiskers) [reply src/test/fuzz/connman.cpp:38]: i remember backporting https://github.com/bitcoin/bitcoin/pull/20828 and for some reason it caused more problems than it solved, forcing me to manually decrement the cases, which was annoying to work with.

File: src/net_processing.cpp (hunk 4 above)
Comment (kittywhiskers) [reply src/net_processing.cpp]: Resolved in latest push

File: src/net_processing.cpp (hunk 3 above)
Comment (kittywhiskers) [reply src/net_processing.cpp]: Resolved in latest push

File: src/init.cpp (hunk 1 above)
Comment (kittywhiskers) [reply src/init.cpp:289]: Resolved backport PR commit fusion

File: test/functional/feature_fee_estimation.py (hunk 2 above)
Comment (kittywhiskers) [reply test/functional/feature_fee_estimation.py:247]: Resolved backport PR commit fusion in latest push

File: src/init.cpp [hunk 6]
```diff
@@ -1769,13 +1769,8 @@ bool AppInitMain(const CoreContext& context, NodeContext& node, interfaces::Bloc
     // Make mempool generally available in the node context. For example the connection manager, wallet, or RPC threads,
```
Comment (knst) [inline src/init.cpp]: nit: comment should be removed in 20222

**Review by knst — APPROVED** [review]
utACK

**Review by knst — DISMISSED** [review]
LGTM

**Review by PastaPastaPasta — APPROVED** [review]
utACK for merging via merge commit

**Review by UdjinM6 — CHANGES_REQUESTED** [review]
pls see below

File: src/net_processing.cpp [hunk 7]
```diff
@@ -1603,7 +1603,7 @@ static void RelayAddress(const CAddress& addr, bool fReachable, CConnman& connma
     assert(nRelayNodes <= best.size());
 
     auto sortfunc = [&best, &hasher, nRelayNodes, addr](CNode* pnode) {
-        if (pnode->nVersion >= CADDR_TIME_VERSION && pnode->IsAddrRelayPeer() && pnode->IsAddrCompatible(addr)) {
+        if (pnode->IsAddrRelayPeer() && pnode->IsAddrCompatible(addr)) {
```
[the remark this answers is not in the document]
Comment (kittywhiskers) [reply src/net_processing.cpp:1657]: resolved in latest push

File: src/net_processing.cpp [hunk 8]
```diff
@@ -938,7 +939,8 @@ void PeerLogicValidation::FinalizeNode(NodeId nodeid, bool& fUpdateConnectionTim
     if (state->fSyncStarted)
         nSyncStarted--;
 
-    if (state->nMisbehavior == 0 && state->fCurrentlyConnected) {
+    if (state->nMisbehavior == 0 && state->fCurrentlyConnected /* && !node.IsBlockOnlyConn() */) {
```
[the remark this answers is not in the document]
Comment (kittywhiskers) [reply src/net_processing.cpp]: resolved in latest push

File: src/net_processing.cpp [hunk 9]
```diff
@@ -986,7 +983,7 @@ void PeerLogicValidation::FinalizeNode(const CNode& node, bool& fUpdateConnectio
     if (state->fSyncStarted)
         nSyncStarted--;
 
-    if (misbehavior == 0 && state->fCurrentlyConnected /* && !node.IsBlockOnlyConn() */) {
+    if (node.fSuccessfullyConnected && misbehavior == 0 /* && !node.IsBlockOnlyConn() */) {
         // Note: we avoid changing visible addrman state for block-relay-only peers
```
[the remark this answers is not in the document]
Comment (kittywhiskers) [reply src/net_processing.cpp]: resolved in latest push

File: src/net_processing.cpp [hunk 10]
```diff
@@ -1007,6 +1002,15 @@ void PeerLogicValidation::FinalizeNode(const CNode& node, bool& fUpdateConnectio
         assert(nPeersWithValidatedDownloads == 0);
         assert(g_outbound_peers_with_protect_from_disconnect == 0);
     }
+    } // cs_main
+
+    if (node.fSuccessfullyConnected && misbehavior == 0 /* && !node.IsBlockOnlyConn() */) {
```
[the remark this answers is not in the document]
Comment (kittywhiskers) [reply src/net_processing.cpp]: resolved in latest push

**Review by UdjinM6 — APPROVED** [review]
utACK

**Review by knst — APPROVED** [review]
LGTM

**Review by PastaPastaPasta — DISMISSED** [review]
utACK for merging via merge commit

**Pull request merged**